Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp

```diff
@@ -390,6 +390,274 @@ ErrorOr<void> InvertCalculationNode::dump(StringBuilder& builder, int indent) co
+    for (auto const& value : m_values)
+        TRY(value->dump(builder, indent + 2));
+    return {};
+}
+
+ErrorOr<NonnullOwnPtr<ClampCalculationNode>> ClampCalculationNode::create(NonnullOwnPtr<CalculationNode> min, NonnullOwnPtr<CalculationNode> center, NonnullOwnPtr<CalculationNode> max)
+{
+    return adopt_nonnull_own_or_enomem(new (nothrow) ClampCalculationNode(move(min), move(center), move(max)));
+}
+
+ClampCalculationNode::ClampCalculationNode(NonnullOwnPtr<CalculationNode> min, NonnullOwnPtr<CalculationNode> center, NonnullOwnPtr<CalculationNode> max)
+    : CalculationNode(Type::Clamp)
+    , m_min_value(move(min))
+    , m_center_value(move(center))
+    , m_max_value(move(max))
+{
+}
+
+ClampCalculationNode::~ClampCalculationNode() = default;
+
+ErrorOr<String> ClampCalculationNode::to_string() const
+{
+    StringBuilder builder;
+    TRY(builder.try_append("clamp("sv));
+    TRY(builder.try_append(TRY(m_min_value->to_string())));
+    TRY(builder.try_append(", "sv));
+    TRY(builder.try_append(TRY(m_center_value->to_string())));
+    TRY(builder.try_append(", "sv));
+    TRY(builder.try_append(TRY(m_max_value->to_string())));
+    TRY(builder.try_append(")"sv));
+    return builder.to_string();
+}
+
+Optional<CalculatedStyleValue::ResolvedType> ClampCalculationNode::resolved_type() const
+{
+    // NOTE: We check during parsing that all values have the same type.
+    return m_min_value->resolved_type();
+}
+
+bool ClampCalculationNode::contains_percentage() const
+{
+    return m_min_value->contains_percentage() || m_center_value->contains_percentage() || m_max_value->contains_percentage();
+}
+
+CalculatedStyleValue::CalculationResult ClampCalculationNode::resolve(Layout::Node const* layout_node, CalculatedStyleValue::PercentageBasis const& percentage_basis) const
+{
+    auto resolve_value = [layout_node](CalculatedStyleValue::CalculationResult::Value value) {
+        return value.visit(
+            [](Number const& number) { return number.value(); },
+            [](Angle const& angle) { return angle.to_degrees(); },
+            [](Frequency const& frequency) { return frequency.to_hertz(); },
+            [layout_node](Length const& length) { return static_cast<float>(length.to_px(*layout_node).value()); },
+            [](Percentage const& percentage) { return percentage.value(); },
+            [](Time const& time) { return time.to_seconds(); });
+    };
+
+    auto min_node = m_min_value->resolve(layout_node, percentage_basis);
+    auto center_node = m_center_value->resolve(layout_node, percentage_basis);
+    auto max_node = m_max_value->resolve(layout_node, percentage_basis);
+
+    auto min_value = resolve_value(min_node.value());
+    auto center_value = resolve_value(center_node.value());
+    auto max_value = resolve_value(max_node.value());
+
+    // NOTE: The value should be returned as "max(MIN, min(VAL, MAX))"
+    auto correct_value = max(min_value, min(center_value, max_value));
```

Comment:
not sure if weʼre here for correctness (whatever questions it might rise)… maybe plain `result_value` or `result` would be enough?